pkg/zeek/escape.go

```diff
@@ -23,6 +25,30 @@ func Escape(data []byte) string {
 	return string(buf)
 }
 
+// EscapeUtf8 does the very same non-standard formatting of mixed-binary strings
+// that zeek does.  There is no way to disambiguate between random binary data
+// and a deliberate utf-8 string so it's left to the "presenation layer" to decide
+// how to format the data.  This would be remedied if a zeek string were a UTF-8
+// string and zeek had a bytes type to represent non-string data.
+func EscapeUtf8(data []byte) string {
+	var scratch [utf8.UTFMax]byte
+	var buf []byte
+	k := 0
+	for len(data) > 0 {
+		c := data[k]
+		r, _ := utf8.DecodeRune(data[k:])
+		if c < 0x20 || c == 0x7f || r == utf8.RuneError {
+			buf = append(buf, '\\', 'x', hexdigits[c>>4], hexdigits[c&0xf])
+			data = data[1:]
+			continue
+		}
+		n := utf8.EncodeRune(scratch[:], r)
```

Comment:
can't we skip re-encoding into scratch and do `buf = append(buf, data[:r])` ?

Reply:
yes, better